Allow multiple loads for same MSRun

lparsons opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

Allow data to be loaded multiple times from the same MSRun across different accucor/isocor files. Duplicate PeakGroups for the same MSRun are now allowed, however.

## Affected Issue Numbers

- Resolves #65
- Resolves #681

## Code Review Notes

This works for the various cases I've tested, but I'm not confident I understand the exception handling rules in the loading classes, so I would appreciate it if @hepcat72 could take a look and see if anything appears way off.

This is a draft PR to facilitate some CI testing and get an early review.

## Checklist

- [x] Changes have been added to the *Unreleased* section in the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*

hepcat72 reviewed (changes requested):
This PR solves a long-standing problem that is a vital issue in supporting one of the primary goals of this project.  I'm glad someone is finally addressing it.  I do have some concerns about the utility of the errors/warnings that are generated (and a minor concern about side-effects of allowing MSRun record associations before we have enough data to definitively distinguish them).

Understanding the principles that went into the error handling development will help when resolving the issues I brought up in this PR.  Without this understanding, it's understandable how the existing error handling could not be entirely clear, so I will provide an explanation.

The overarching goal was to reduce the time and effort it takes for the user/curator to resolve errors resulting from loading.  Thus, I established the following guidelines when implementing the exception handling strategy.  I hadn't typed them all out discretely before, so this list is a work-in-progress.  Perhaps they should be documented in some form other than a PR Review description...

Errors should be:

- **Succinct** (Summarize repetitive errors.  Repetitive errors can "hide" other important errors, which makes them easy to miss.  It also lengthens the amount of time and attempts it takes to resolve them because you have to carefully scrutinize large amounts of text to catch the different error cases that exist.)
- **Informative** (In order for the user to be able to decide what to do about the error, they need context.  If the error is about a duplicate record, they should be provided with information about the duplicate (e.g. what file it came from).)
- **Distinct** (If there are different causes or solutions to the same error, present them as different errors.  The error class name should make resolving them rote.  I.e. If the user has seen error class name before, they should be able to apply the same resolution as they have applied in the past (if we have enough information to distinguish them).)
- Useful (Errors that are the side-effects of other earlier errors, they should be skipped/suppressed.)
- **Actionable** (In order for a user to be able to act on the error, they need to be able to find the data that is causing the error.  In the case of spreadsheets, they need to know the file, row(s), and column(s) involved.  Often, the default `ValidationError` or other django-issued error provides no useful information, but we *do* have information that we could provide that would drastically reduce the amount of effort involved in figuring out what the real problem is, thus such errors should be repackaged to include this information.  And if a row or column identifier is not unique, row/col numbers can help speed up the identification of the data involved.)
- **Readable** (Key pieces of data should be either bulleted or indented so that it's easy to visually scan.)
- **Complete** (Find as many errors as is possible to reduce the number of loading attempts.)
- **Mapped to a resolution** (The number of possible resolutions per error type should be minimized as much as possible.  The error should suggest possible resolutions and if there's more than one, it should provide information necessary to know which resolution should be applied.)

lparsons commented on DataRepo/tests/loading/test_load_accucor_msruns.py:
```diff
@@ -197,10 +198,10 @@ def test_existing_msruns(self):
         aes = ar.exception
-        self.assertEqual(1, aes.num_errors)
-        self.assertEqual(MSRunAlreadyLoadedOrNotUnique, type(aes.exceptions[0]))
+        self.assertEqual(28, aes.num_errors)
+        self.assertEqual(DuplicatePeakGroup, type(aes.exceptions[0]))
```
Yeah, that would be a lot nicer for the user. Might make sense to have one error per AccuCor file that indicates the number of duplicate peaks.

* [ ] Consolidate multiple DuplicatePeakGroup errors for each peak annotation file (accucor/isocor)

lparsons commented on DataRepo/utils/accucor_data_loader.py:
```diff
@@ -725,41 +725,10 @@ def load_data(self):
                     print(
                         f"No cache exists for animal {msrun.sample.animal.id} linked to Sample {msrun.sample.id}"
                     )
-            except ValidationError as ve:
-                vestr = str(ve)
-                if (
-                    "Mass spectrometry run with this Researcher, Date, Protocol and Sample already exists"
-                    in vestr
-                ):
-                    # PR REVIEW NOTE: The file for the peak_group_set should probably be a field in the MSRun table and
-                    # be included in the UniqueConstraint.  Then we wouldn't have to tweak the dates.
-                    existing_file = (
-                        MSRun.objects.get(**msrun_dict)
-                        .peak_groups.first()
-                        .peak_group_set.filename
-                    )
-                    self.existing_msruns[existing_file].append(sample_name)
-                else:
-                    self.aggregated_errors_object.buffer_error(ve)
-                continue
```
Certainly having mzXML files will help us uniquely identify samples, though I still believe the key is identifying the _animals_ correctly. It seems highly doubtful that there would be multiple liver samples from the same animal, or at least if there were, they would be uniquely named. Sample names really only need to be unique among those samples belonging to same animal. Animal names do need to be unique across the database, at least if we are allowing them to be part of multiple studies.

I'll add the comment, but I'm not sure what the issue would be at the moment.

lparsons replied:
I'm looking more carefully at the logic to get a sample record in this file, and I'm realizing the issue is that we don't know which Animal the Sample came from because we don't have the context of the animals and samples sheet provided with the accucor file. So yeah, this is a bit bigger of a problem the way we are loading things now. We should consider this a bit when we rework the submission stuff. Would be nice to make sure we are working off the correct set of animals each time (not sure that's feasible, but would be nice).

lparsons replied:
Thinking about it more, as long as people are submitting animal and sample sheets along with the accucor files, we should catch these problems when we process those files. If there was a sample with the same name, but a different animal, it would error out then.

hepcat72 replied:
> as long as people are submitting animal and sample sheets along with the accucor files, we should catch these problems when we process those files

Just started this second round review this morning, so I have yet to look at the code changes you would have needed to make to make this possible, but on the off chance you're asserting that existing code will catch this issue (associating samples from the accucor files to the wrong sample associated with a different animal), I don't think that's true.  However, that said, I don't think it's as big of an issue as I originally thought.

But first, the sample data *must* be loaded in order to be able to load the associated accucor/isocorr files.  Otherwise, you would get errors about missing samples.  Accucor data is always loaded after sample data.  It wouldn't work without it, but it doesn't solve this issue.

Regarding the creation of the `MSRun` record, there is no checking of the validity of the sample name with resect to the contents of the sample sheet.  It only uses the sample record it got from the database: `Sample.objects.get(name=prefixed_sample_name)`.  There's no direct comparison between the sample headers in the accucor file and the contents of the sample column from the sample sheet.  They could be anything.  The only thing it checks is if the sample exists in the database.  If a user has 2 accucor files from different runs on the same day (which has occurred before), there are a few scenarios I can see happening:

1. **Same sample name** as pre-existing Sample record
   1. **Same researcher, day, protocol** as existing MSRun record
      1. The existing sample record is **actually a different sample** (same or different animal - doesn't matter, it's still wrong)
         - **This is what we currently, but will no longer catch as a result of this PR, though I can't say whether it has ever happened - probably hasn't**
      2. The existing sample record is actually the same sample
         - This is where we currently have to fudge the date (which has been [done, commented on, and approved](https://github.com/PrincetonUniversity/tracebase-rabinowitz-data/blob/272969987d58e99c980e2ce7dfa49315792a857d/cold_exposure/CHANGES.md?plain=1#L142) in the past)
   2. Different researcher, day, protocol - no existing MSRun record
      1. different sample (same or different animal - doesn't matter, it's still wrong)
         - This is currently a hole where we may have incorrectly associated samples with the wrong animal in the past - it is not caught, and the code in this PR still leaves this issue extant.  It can only happen if the sample name in the accucor file is not in the sample sheet, but exists in the database due to a previous load.
      2. same sample
         - This is what this PR enables
2. *Different sample name* as any pre-existing Sample record
   - This always results in a missing sample error

The reason item `1.1.1.` can happen is because there's no check between the values in the samplesheet and accucor **files**.  There would have to be a discrepancy in the files of the submission where 2 **different** samples have the same name in 2 different accucor files.  They could be the same MSRun or 2 different MSRuns, but that doesn't matter - we have no way to distinguish them if they were different.

I feel like I created an issue to deal with this before where I proposed that we add a way (e.g. an option) to supply the sample names from the sample sheet to the accucor load, but I could not find it in a search, so perhaps we can add that to #684.

I was conflating `1.1.1.` with `1.2.1.`, which I realized we're at current risk of silently making a bad association, and this PR doesn't prevent that either.  We would be able to detect this issue by doing a database query to look for animals associated with MSRn records with different researchers.  But since 2 things would have to happen: *forget to include the sample name from the accucor file in the sample sheet* and *match an existing sample name from a different study*, it seems pretty unlikely to happen.

lparsons commented on DataRepo/utils/accucor_data_loader.py:
```diff
@@ -805,14 +774,35 @@ def load_data(self):
+                                "Peak group with this Name and Msrun already exists"
+                                in vestr
+                            ):
+                                self.aggregated_errors_object.buffer_exception(
```
Thanks for the guidance in consolidating the errors, that will help.

I'm not sure what difference it makes if the entire file contains "duplicate peaks" or not. If it is the same file, the name would likely match, thus the PeakGroupSet would match. In coming iterations we'll store that and have a checksum too. In that case, it would be nice to just skip it, though ideally we'd check to see that the data is all the same. I do think it would be nice to have a summary per file of PeakGroups added, skipped, etc. but I'm not sure that's in scope here, especially since we have some larger changes coming shortly. In fact, I suspect we'll make those changes before we get more than one or two new studies to load.

Good catch about the `IntegrityError`, that should be caught here.

* [ ] Handle `IntegrityError` when inserting `PeakGroup` and repackage as `ConflictingValueError`.

lparsons replied:
Actually, we're not doing a `get_or_create` here (though perhaps we should), so I don't believe there will be any `IntegrityError` raised here. Instead, we are calling [`full_clean()`](https://docs.djangoproject.com/en/3.2/ref/models/instances/#django.db.models.Model.full_clean) which will raise a `ValidationError` instead, which we are catching.

hepcat72 replied:
In either case, whatever type exception is raised, there are 2 possible error causes:

1. All the record data is the same in an existing record in the database (I believe this is what you're handling)
2. The PeakGroup's name and MSRun value combo exists, but one or both of the other fields, (formula or peakgroupset), differs.

These are 2 circumstances that require different solutions.  The data needed for resolving the second error case should be used to create a `ConflictingValueError`, which packages all of the useful data for solving the second case.  I suggest writing a test for this case and see what you get, then repackage that as a `ConflictingValueError`.  If both cases result in the same exact `ValidationError` exception, the case will have to be caught in code, not via raised exception identification.

lparsons replied:
> 2\. The PeakGroup's name and MSRun value combo exists, but one or both of the other fields, (formula or peakgroupset), differs.
...
> If both cases result in the same exact ValidationError exception, the case will have to be caught in code, not via raised exception identification.

Both seem to do the same thing since we are not attempting to lookup the existing PeakGroup. I realize that this means we would raise and error even when attempting to load the exact same data, but that has been the case since before this PR (we would have raised a duplicate MSRun exception before). I believe that is out of scope, but I'll create and issue (which I suggest we resolve after addressing mzXML file loading).

lparsons replied:
So, I'm a bit confused, and perhaps I'm still misunderstanding things. We are talking about inserting a `PeakGroup` record here. If the `msrun` and the `name` are the same, then it's a duplicate and we won't insert it regardless of the other values. The only other values to check are, the `formula` and the `peak_group_set`. The `formula` is determined by the compound lookup done from the `name` and the `peak_group_set` is the file it was loaded from. Regardless of those values, the PeakGroup insertion won't be allowed.

Since there is only a create here, there is no chance for an `IntegrityError` to be raised, only a `ValidationError` from the `full_clean()` call. In fact, the test is to attempt to load the PeakGroup from a file with a different name (different `peak_group_set`) which is done in `test_duplicate_compounds_one_msrun`. Is there some other case we should be testing?

In theory, it would be nice to lookup the existing peak data for the PeakGroup and check it against the PeakData in the file being loaded and skip over it if it's the same, but I think that is a bit out of scope here.

lparsons commented:
@fkang-pu and @jcmatese You are welcome to review this and just take a look. 

@hepcat72 I believe I have addressed the issues either in a comment or in the code (or both).

hepcat72 reviewed (changes requested):
I have 2 new blocking issues related to incorrect `assertTrue` calls.  There are a few other non-blocking issues that could be ignored as you see fit. But regarding the one blocking issue I previously raised, I still feel like there should be a distinction between duplicate records and conflicting values. I'm willing to consider it non-blocking for the sake of progress, but I generally feel like the error should simply be transparent about the differences in data that is deemed as "duplicate" if not all data in the record is the same.

There was one other thing I realized during this review, and that is the potential for incorrect and silently loaded associations of samples with the wrong animal - which is currently in main and persists in this PR.  My [comment](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/682#discussion_r1204645611) makes a suggestion WRT #684.  It's in the outline in position `1.2.1.`.

lparsons commented:
@hepcat72 I've changed the two `assertTrue` calls to `assertEquals` calls, good catch there. Thanks.

I spent some time trying to create a situation where we had duplicate vs conflicting data for PeakGroups and came to conclusion that we should do a few things to deal with the issues you raised. The issues as I understand them are related to lookups of complex objects by names, which could result in incorrect associations being made.

1. Sample names in the peak annotation files might refer to an incorrect, previously loaded sample of the same name. The best solution to that so far is to require that peak annotation files always have an accompanying sample/animal sheet and we ensure the sample/animals load/match correctly. In addition, ensure that the peak annotation sample names match one in the provided sample sheet. That is documented in issue #684.
2. Compounds in peak annotation files may refer to an incorrect existing compound of the same name. In many cases, the peak annotation file contains a formula for the compound. We could ensure that the formula supplied in the peak annotation file is the same as the compound retrieved from the database of the same name. If we agree, I'll create an issue for this. This would handle the case of mismatched formula data in a PeakGroup record.
3. Peak data (and peak groups) that were already loaded could in theory be skipped if the data matched exactly. To implement any meaningful check, I suspect we would need to refactor how PeakGroups and PeakData are loaded so that we can check the associated values when loading. This would allow us to either skip loading a PeakGroup where all the associated PeakData matched or report and error of a duplicate PeakGroup if the data did not match. We can also create an issue for this.

Right now, there are two values for PeakGroups that are not part of the unique key that we could check. One is `formula`, which should really be checked when loading the compounds. The other is `peak_group_set` (which is the peak annotation file name), which must be unique at the moment due to how PeakGroupSet records are created in the loading code already. They are always created, not retrieved, thus any duplicate results in an error before we attempt to load PeakGroups. Thus, I think that solving those issues is outside the scope of this PR.

lparsons commented:
I have refactored some of the peakgroup handling code to make it a bit more modular and make testing a bit simpler.

One concern I have is the handling of multiple `DuplicatePeakGroup` and `ConflictingValueError` exceptions. This code currently collects them in a list and packages them up in a new Exception type. I'm not sure that is particularly elegant or scalable to other error types. It does seem that `AggregatedErrors` has some functionality to consolidate multiple instances of the same exception, so perhaps each duplicate / conflict should just be added to the `AggregatedErrors` object?

hepcat72 commented:
> I have refactored some of the peakgroup handling code to make it a bit more modular and make testing a bit simpler.
> 
> One concern I have is the handling of multiple `DuplicatePeakGroup` and `ConflictingValueError` exceptions. This code currently collects them in a list and packages them up in a new Exception type. I'm not sure that is particularly elegant or scalable to other error types. It does seem that `AggregatedErrors` has some functionality to consolidate multiple instances of the same exception, so perhaps each duplicate / conflict should just be added to the `AggregatedErrors` object?

I started out simply adding errors to AggregateErrors, but the output in my opinion was overwhelming and difficult to navigate.  Grouping and summarizing multiple errors of the same type was a vast improvement.  It keeps errors from mixing, which also makes it less likely you'll miss something.

I recommend, when you create or reorganize exceptions, to run the test data through the validate interface to see how well they display.  And the dataset which I used during development had tons of errors (cold exposure), which highlighted the need for consolidation and summarization.  I kept that erroneous data around on my machine to use for more testing.

I do agree that the code could be more modular/extensible and could be improved, but I think that the same principles about navigability/consumability should be applied to any refactor.

jcmatese reviewed (approved):
Looks ok to me...

lparsons commented on DataRepo/utils/accucor_data_loader.py:
```diff
@@ -695,22 +817,29 @@ def load_data(self):
-                msrun.save()
+                # TODO This relies on sample name lookup (see issue #684)
+                # https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/684
+                """This also relies on accurate msrun information (researcher and date)
+                Including mzXML files with accucor files will help ensure accurate msrun
+                lookup since we will have checksums for the mzXML files and those are
+                always associated with one MSRun record
+                """
```
I agree that other options for defining MSRun would be good to consider. Just to clarify, this PR allows one to load multiple files from the same MSRun, which I believe we always intended to support.

hepcat72 reviewed (changes requested):
One blocking issue (that should be trivial to fix), but otherwise - this is great.  It exactly covers the issue we had that long discussion about from the first review.  You nailed it.

And if you go with my suggestion from the blocking issue (and it doesn't prompt further issues/discussion), then you can consider this PR approved, and merge as you see fit.

lparsons commented on DataRepo/tests/loading/test_load_accucor_msruns.py:
```diff
@@ -174,39 +190,70 @@ def test_record_missing_compound(self):
+            "compounds": peak_group.compounds,
+        }
+
+        with self.assertRaises(DuplicatePeakGroup):
+            adl.insert_peak_group(
+                peak_group_attrs,
+                msrun=peak_group.msrun,
+                peak_group_set=peak_group.peak_group_set,
+            )
```
I don't think skipping duplicate peak groups would be wise without further testing. Also, it is impossible to load a duplicate peak group since duplicate PeakGroupSet is not allowed. IMO, The functionality should be more clearly defined and implemented in a separate effort. I also think it should be tacked after we incorporate mzXML file storage.

hepcat72 replied:
I should note that since you made the change I suggested WRT `is_fatal=self.validate` and `is_error=False`, it means that in the curator/loading context, only a warning will be printed.  In the user validation interface context, it will still be a warning, but it is raised as an exception, and thus presented to the user as a yellow warning message. If you didn't realize that that's what that change would mean, and still want it to prevent loading in the curator/loading context, then you should set `is_fatal=True`.

Personally, I think that the current state of error handling for this type of exception is ideal and will mark this issue as resolved.

lparsons commented on DataRepo/utils/exceptions.py:
```diff
@@ -940,10 +963,10 @@ def __init__(
-                f"[{str(model_to_dict(rec))}]:\n"
-                f"\tdatabase: [{existing_value}]\n"
-                f"\tfile: [{differing_value}]"
+                f"Conflicting {consistent_field} field values encountered {rowmsg}in {type(rec).__name__} record "
+                f"{str(model_to_dict(rec))}:\n"
+                f"\tdatabase: {existing_value}\n"
+                f"\tfile: {differing_value}"
```
I made this change because the output was very difficult to parse, containing duplicate delimiters. Perhaps that is not always the case though, so I've reverted the change.

lparsons replied:
Sorry, I meant readability. It was very difficult to read with the extra delimiters, at least to me, but I'm fine leaving them.

hepcat72 replied:
You could change the delimiter, or conditionally display the delimiters (if the value is an empty string, `None`, or has white space on one of the ends... though changing the delimiter would be easier.

lparsons commented on DataRepo/utils/accucor_data_loader.py:
```diff
@@ -1046,8 +1101,13 @@ def load_data(self):
                     duplicate_peak_groups=duplicate_peak_groups,
                 ),
-                is_fatal=True,
-                is_error=not self.validate,
```
OK, I'll change those values. My intent was to disallow loading of duplicates in this case since we have not done enough checking downstream on peak data values, etc. but since this error cannot be reached in practice, I'm OK with changing it.

hepcat72 reviewed (approved):
Nothing blocking.  This is good work.

If the ci tests pass, I say merge it.

hepcat72 commented on DataRepo/utils/exceptions.py:
```diff
@@ -86,37 +86,18 @@ def __init__(self, missing, animal_hdr="animal", message=None):
+            f"\tms_run: {ms_run}\n"
+            f"\tsample: {sample}\n"
+            f"\tcompound: {compound}\n"
         )
```
I think you did this.  I will check your todo item off and resolve.
